renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -420,9 +435,749 @@ ShaderDebugTrace D3D12Replay::DebugVertex(uint32_t eventId, uint32_t vertid, uin
+
+  // Create the descriptor table for our UAV
+  D3D12_DESCRIPTOR_RANGE1 descRange;
+  descRange.RangeType = D3D12_DESCRIPTOR_RANGE_TYPE_UAV;
+  descRange.NumDescriptors = 1;
+  descRange.BaseShaderRegister = 0;
+  descRange.RegisterSpace = 0;
+  descRange.Flags = D3D12_DESCRIPTOR_RANGE_FLAG_NONE;
+  descRange.OffsetInDescriptorsFromTableStart = 0;
+
+  modsig.Parameters.push_back(D3D12RootSignatureParameter());
+  D3D12RootSignatureParameter &param = modsig.Parameters.back();
+  param.ParameterType = D3D12_ROOT_PARAMETER_TYPE_DESCRIPTOR_TABLE;
+  param.ShaderVisibility = D3D12_SHADER_VISIBILITY_PIXEL;
+  param.DescriptorTable.NumDescriptorRanges = 1;
+  param.DescriptorTable.pDescriptorRanges = &descRange;
+
+  uint32_t sigElem = uint32_t(modsig.Parameters.size() - 1);
+
+  // Create the root signature for gathering initial pixel shader values
+  ID3DBlob *root = m_pDevice->GetShaderCache()->MakeRootSig(modsig);
+  ID3D12RootSignature *pRootSignature = NULL;
+  hr = m_pDevice->CreateRootSignature(0, root->GetBufferPointer(), root->GetBufferSize(),
+                                      IID_PPV_ARGS(&pRootSignature));
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create root signature for pixel shader debugging HRESULT: %s",
+           ToStr(hr).c_str());
+    SAFE_RELEASE(root);
+    SAFE_RELEASE(psBlob);
+    SAFE_RELEASE(pInitialValuesBuffer);
+    SAFE_RELEASE(pInitialValuesStaging);
+    return empty;
+  }
+  SAFE_RELEASE(root);
+
+  WrappedID3D12PipelineState *origPSO =
+      m_pDevice->GetResourceManager()->GetCurrentAs<WrappedID3D12PipelineState>(rs.pipe);
+
+  RDCASSERT(origPSO->IsGraphics());
+
+  D3D12_EXPANDED_PIPELINE_STATE_STREAM_DESC pipeDesc;
+  origPSO->Fill(pipeDesc);
+
+  // All PSO state is the same as the event's, except for the pixel shader and root signature
+  pipeDesc.PS.BytecodeLength = psBlob->GetBufferSize();
+  pipeDesc.PS.pShaderBytecode = psBlob->GetBufferPointer();
+  pipeDesc.pRootSignature = pRootSignature;
+
+  ID3D12PipelineState *initialPso = NULL;
+  hr = m_pDevice->CreatePipeState(pipeDesc, &initialPso);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create PSO for pixel shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(psBlob);
+    SAFE_RELEASE(pInitialValuesBuffer);
+    SAFE_RELEASE(pInitialValuesStaging);
+    SAFE_RELEASE(pRootSignature);
+    return empty;
+  }
+
+  // Add the descriptor for our UAV
+  std::set<ResourceId> copiedHeaps;
+  PortableHandle shaderDebugUav = ToPortableHandle(GetDebugManager()->GetCPUHandle(SHADER_DEBUG_UAV));
+  AddDebugDescriptorToRenderState(m_pDevice, rs, shaderDebugUav,
+                                  D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV, sigElem, copiedHeaps);
+
+  // Set the PSO and root signature
+  rs.pipe = GetResID(initialPso);
+  rs.graphics.rootsig = GetResID(pRootSignature);
+
+  ID3D12GraphicsCommandListX *cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
+  rs.ApplyState(m_pDevice, cmdList);
```

Comment:
Do we need to apply the whole state here? I think we just need to set the CBV/SRV/UAV heap, right?

Reply:
We should need to set the PSO, root sig, and the new parameter as well as the heap, right? With a test to simply remove the ApplyState call, things still _appear_ to work, but I'm left confused about where that state is getting applied to be able to capture the PS inputs. Does the render state struct get applied elsewhere?

Reply:
This command list is only calling `ClearUnorderedAccessViewUint` before you close it on line 766 below and then submit it, so we don't need any of those things I don't think. Applying the whole state probably isn't a big deal but I'm not sure how expensive that is and if we aren't doing any draws then we're just as well only setting what we need.

The draw that captures the PS inputs is done below in a separate command list inside `ReplayLog`. That function uses the current render state when you're doing an `eReplay_OnlyDraw` replay, so in that case it's sufficient to just change the contents of `rs` (which is a reference pointing to the current `m_RenderState` which gets applied on replay).

Reply:
Thanks for the context of how m_RenderState is used. I tested it further and only setting the heaps (if a new one was added) does suffice here.